Bugfix and optimizations for scans that select no columns

JoshRosen opened the pull request:
If a query resulted in a Redshift scan that selected no columns (such as calling `.count()` on a DataFrame created from a Redshift table), then `spark-redshift` would generate an invalid `UNLOAD` query.

This patch fixes this issue by converting these Redshift scans into `SELECT count(*)` queries, bypassing the unload. This has the added benefit of offering a large performance improvement for such queries.

JoshRosen commented on src/main/scala/com/databricks/spark/redshift/RedshiftRelation.scala:
```diff
@@ -49,50 +49,74 @@ private[redshift] case class RedshiftRelation(
+        classOf[java.lang.Long], classOf[Array[String]], hadoopConf)
+      val prunedSchema = pruneSchema(schema, requiredColumns)
+      rdd.values.mapPartitions { iter =>
+        val converter: Array[String] => Row = Conversions.createRowConverter(prunedSchema)
```
Here, I took the opportunity to do a bit of minor performance optimization by re-using the row converter rather than creating a new one for each row. There are many other inefficiencies / opportunities for optimization in the converter itself, but I'll address those in a separate patch.

JoshRosen commented on src/main/scala/com/databricks/spark/redshift/RedshiftRelation.scala:
```diff
@@ -49,50 +49,74 @@ private[redshift] case class RedshiftRelation(
     s"UNLOAD ('$query') TO '$fixedUrl' WITH CREDENTIALS '$credsString' ESCAPE ALLOWOVERWRITE"
   }
 
-  private def makeRdd(schema: StructType): RDD[Row] = {
```
Here, I just inlined the bodies of these functions after some refactoring.  I think the new code is slightly easier to follow.

marmbrus commented on src/it/scala/com/databricks/spark/redshift/RedshiftIntegrationSuite.scala:
```diff
@@ -240,6 +240,10 @@ class RedshiftIntegrationSuite
   }
 
+  test("count() on DataFrame created from a Redshift table") {
+    assert(sqlContext.sql("select * from test_table").count() === TestUtils.expectedData.length)
```
This is minor, but even for things like this I'd consider using `checkAnswer` and explicit `COUNT(*)`, just cause when things do go wrong the query plans can be insightful.

JoshRosen replied:
Good idea; will update.

marmbrus commented on src/main/scala/com/databricks/spark/redshift/RedshiftRelation.scala:
```diff
@@ -49,50 +49,74 @@ private[redshift] case class RedshiftRelation(
-    val statement = conn.prepareStatement(unloadSql)
+  private val tableNameOrSubquery: String = {
+    val unescaped = params.query.map(q => s"($q)").orElse(params.table).get
+    unescaped.replace("'", "\\'")
```
Why is this escaping needed?

JoshRosen replied:
This is a carryover from old code, but I should have documented it there. In a nutshell, the issue is that the subquery / table name that is passed to the UNLOAD command ends up being enclosed in single quotes, so any single quotes within that string need to be escaped another time to account for the additional layer of outer single quotes.  I'll add an inline comment to explain this, since it's not obvious at first glance.

JoshRosen replied:
Actually, on closer inspection I think that this is actually a mistake when we're handling an empty column query, since in those cases we won't end up adding the extra layer of single quotes. I'm going to add a test to expose this bug and will refactor to fix this.

marmbrus commented:
LGTM

JoshRosen commented on src/main/scala/com/databricks/spark/redshift/RedshiftRelation.scala:
```diff
@@ -49,50 +49,74 @@ private[redshift] case class RedshiftRelation(
+        val results = conn.prepareStatement(countQuery).executeQuery()
+        if (results.next()) {
+          val numRows = results.getLong(1)
+          val parallelism = sqlContext.getConf("spark.sql.shuffle.partitions", "200").toInt
```
Whoops, this is unused. I'll update to pass it into `parallelize`.

JoshRosen commented:
Added a regression test for the subquery bug, added more comments, fixed the bug, and updated to use checkAnswer, so I'm going to merge this. Thanks for the review.
